### ISSUE TITLE
The Fluent IconServer.cs is changing the HttpClient properties

### ISSUE THREAD

**Issue by eeegs** [microsoft/fast-blazor#215]
# 🐛 Bug Report

If I am loading data from the server and use one of the overrides OnAfterInitializeAsync or OnAfterRenderAsync my request seems to get fired first so the modification of the HttpClient.BaseAddress causes this exception below when a FluentIcon is on the page.

## 🤔 Expected Behavior

Adding an Icon to the page should not interfere with my data requests.

## 😯 Current Behavior

Microsoft.AspNetCore.Components.WebAssembly.Rendering.WebAssemblyRenderer[100]
      Unhandled exception rendering component: This instance has already started one or more requests. Properties can only be modified before sending the first request.
System.InvalidOperationException: This instance has already started one or more requests. Properties can only be modified before sending the first request.
   at System.Net.Http.HttpClient.CheckDisposedOrStarted()
   at System.Net.Http.HttpClient.set_BaseAddress(Uri value)
   at Microsoft.Fast.Components.FluentUI.IconService..ctor(HttpClient httpClient, NavigationManager navigationManager) in /_/src/Microsoft.Fast.Components.FluentUI/IconService.cs:line 15

## 💁 Possible Solution

Don't adjust the HttpClient.BaseAddress in your code.
Or better yet don't use a service to get the icon.  Generate the correct HTML.

## 🌍 Your Environment

Latest of everything, Edge, Windows.

**Comment by vnbaaij** [comment]
> Or better yet don't use a service to get the icon. Generate the correct HTML.

Can you enlighten me on what you mean by that? If you think there is a better way to get the icons, I'm all ears!

**Comment by Notheisz57** [comment]
Maybe their issue is that they expected the `IconService` to be _isolated_ from their user-code?
It appears the `IconService` has a constructor parameter `HttpClient`, presumably fulfilled by DI.
Maybe DI is providing the `IconService` with a singleton `HttpClient` that the user intended to use?

I think the _Getting Started_ says to call `Services.AddHttpClient()`, which should configure/provide a `IHttpClientFactory` via DI.
Maybe changing the `IconService` constructor to take an `IHttpClientFactory` parameter instead of the `HttpClient`, then using that factory to create an instance of `HttpClient` within the constructor would solve the issue?

This is just my quick glance though.

**Comment by eeegs** [comment]
Sorry I haven't responded sooner.

I have been doing some work that has improved my understanding of the use of `HttpClient`. I think Notheisz is close.

In my code the `HttpClient` is the OOTB: `builder.Services.AddScoped(sp => new HttpClient { BaseAddress = new Uri(builder.HostEnvironment.BaseAddress) });`.

The problem isn't that the DI is giving us the same instance per se, it is that the `BaseAddress` in being set in the `IconService` after it is set at startup.  That is a no-no.  If the icon service needs to set a different `BaseAddress` you'll have to use a 'private' `HttpClient`.  The main application must be free to set what it considers to be the `BaseAddress`.

So, either the factory as suggest by Notheisz57 or _new_ one up.

But as to my initial post, I must be missing something obvious, but what is the `IconService` using code to get the icon.  Why is it needing a `HttpClient` and the `NavigationManager` at all?  I also note that you are just setting the `BaseAddress` to the existing `BaseAddress`, which I would have thought is redundant.  It is the act of setting that property that is causing the problem, because as per the original error, sometimes the main code has already launched a data request, and from then on, the `HttpClient`'s properties can't be updated.

I'm going to pull the code to see if I can get a better understanding of why the need for the `HttpClient`.

Regards,

Scott

**Comment by eeegs** [comment]
Now that I've looked properly... ;)

The `IconService` seems to be redundant.  It only being used to hand over a the `HttpClient` anyway.  Why not just inject a `HttpClient `directly?

Pull request coming.

**Comment by eeegs** [comment]
I didn't create a pull request, but here is my recommendation for IconService.cs:

```
using Microsoft.AspNetCore.Components;

namespace Microsoft.Fast.Components.FluentUI;

public class IconService
{
    public HttpClient HttpClient { get; }
    public NavigationManager NavigationManager { get; }

    public IconService(HttpClient httpClient, NavigationManager navigationManager)
    {
        HttpClient = httpClient;
        NavigationManager = navigationManager;
        HttpClient.BaseAddress ??= new Uri(NavigationManager.BaseUri);
    }

}

```

**Comment by vnbaaij** [comment]
I added the service because the icons are not available client side in Blazor Web Assembly. They are plain static resources that need to be downloaded from the server when needed and I need an HttpClient to do that. From an earlier blogpost:
In other libraries you often find icons being included by means of fonts. This comes with the disadvantage that the whole font needs to be downloaded even if you are only using 1 icon. I therefore stepped away from using that method in this library and opted for taking the SVG route. Because the icons are SVG files that are stored in the wwwroot folder in a RCL (Razor Class Library), they are being treated like ordinary static files on the server. They don’t get downloaded until requested and only the icons you are actually using will be downloaded.

The HttpClient BaseAddress was empty sometimes so that is why this construct exists. The null coalition seems like a good fix. Thanks. I'll test it and implement if all checks out good.

**Comment by Notheisz57** [comment]
Apologies, I was generalizing when I referred to the `HttpClient` as being a singleton service; I was not saying that it was being registered as a singleton service.
BaseUri is an instance specific property, so unless you both use the same one, you should never run into this issue, no?

I could be mistaken, but I think this is likely caused by [Scoped services in WebAssembly actually being singletons (see "Service lifetime" section)](https://docs.microsoft.com/en-us/aspnet/core/blazor/fundamentals/dependency-injection?view=aspnetcore-6.0#service-lifetime)

**Comment by eeegs** [comment]
Hi Vincent.

“Sometimes”, depend of the type of app, a hosted wasm, calling back to the service for data certainly sets the baseaddress by default.

BTW, I have now had a proper look and saw what you are doing with the svg files, and like your approach.  Thanks for your effort.

**Comment by vnbaaij** [comment]
Fixed in #220. Will be part of 1.5.3
